File: 30 Days of Code/23 - Day 22 - Binary Search Trees.py
# ========================
#       Information
# ========================

# Direct Link: https://www.hackerrank.com/challenges/30-binary-search-trees/problem
# Difficulty: Easy
# Max Score: 30
# Language: Python

# ========================
#         Solution
# ========================

class Node:
    def __init__(self, data):
        self.right = self.left = None
        self.data = data
class Solution:
    def insert(self, root, data):
        if root == None:
            return Node(data)
        else:
            if data <= root.data:
                cur = self.insert(root.left, data)
                root.left = cur
            else:
                cur = self.insert(root.right, data)
                root.right = cur
        return root

    def getHeight(self, root):
        # Write your code here
        if root != None:
            l_height = 0
            r_height = 0
            l_height = self.getHeight(root.left) + 1
            r_height = self.getHeight(root.right) + 1
            return max(l_height, r_height)
        else:
            return -1

T = int(input())
myTree = Solution()
root = None
for i in range(T):
    data = int(input())
    root = myTree.insert(root, data)
height = myTree.getHeight(root)
print(height)

File: 30 Days of Code/C#/23 - Day 22 - Binary Search Trees.cs
// ========================
//       Information
// ========================

// Direct Link: https://www.hackerrank.com/challenges/30-binary-search-trees/problem
// Difficulty: Easy
// Max Score: 30
// Language: C#

// ========================
//         Solution
// ========================

using System;

class Node {
	public Node left,
	right;
	public int data;
	public Node(int data) {
		this.data = data;
		left = right = null;
	}
}

class Solution {

	static int getHeight(Node root) {
		//Write your code here
		if (root == null) {
			return - 1;
		}

		return 1 + Math.Max(getHeight(root.left), getHeight(root.right));
	}

	static Node insert(Node root, int data) {
		if (root == null) {
			return new Node(data);
		}
		else {
			Node cur;
			if (data <= root.data) {
				cur = insert(root.left, data);
				root.left = cur;
			}
			else {
				cur = insert(root.right, data);
				root.right = cur;
			}
			return root;
		}
	}
    
	static void Main(String[] args) {
		Node root = null;
		int T = Int32.Parse(Console.ReadLine());
		while (T-->0) {
			int data = Int32.Parse(Console.ReadLine());
			root = insert(root, data);
		}
		int height = getHeight(root);
		Console.WriteLine(height);
	}
}
